pkg/api/customization/multiclusterapp/multi_cluster_app.go

```diff
@@ -1,13 +1,77 @@
 package multiclusterapp
 
 import (
-	"github.com/rancher/norman/api/handler"
+	"fmt"
+	"net/http"
+
+	"github.com/rancher/norman/api/access"
 	"github.com/rancher/norman/types"
+	"github.com/rancher/norman/types/values"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	mgmtschema "github.com/rancher/types/apis/management.cattle.io/v3/schema"
+	v3client "github.com/rancher/types/client/management/v3"
 	"github.com/sirupsen/logrus"
 )
 
-func ListHandler(request *types.APIContext, next types.RequestHandler) error {
-	logrus.Debugf("Multi cluster app list handler called")
+const (
+	impersonateUserHeader = "Impersonate-User"
+)
+
+type Wrapper struct {
+	MultiClusterAppLister v3.MultiClusterAppLister
+}
+
+func (w Wrapper) Validator(request *types.APIContext, schema *types.Schema, data map[string]interface{}) error {
+	if request.Method != http.MethodPut && request.Method != http.MethodPost {
+		return nil
+	}
+
+	targetProjects := make(map[string]bool)
+	// check if the creator of multiclusterapp and all members can access all target projects
+	targetMapSlice, found := values.GetSlice(data, "targets")
+	if !found {
+		return fmt.Errorf("no target projects provided")
+	}
+	for _, t := range targetMapSlice {
+		projectID, ok := t["projectId"].(string)
+		if !ok {
+			continue
+		}
+		if !targetProjects[projectID] {
+			targetProjects[projectID] = true
+		}
+	}
+
+	membersToValidate := make(map[string]bool)
+	creatorID := request.Request.Header.Get(impersonateUserHeader)
+	membersToValidate[creatorID] = true
+	membersMapSlice, found := values.GetSlice(data, "members")
+	if found {
+		for _, m := range membersMapSlice {
+			userID, ok := m["userId"].(string)
+			if !ok {
+				continue
+			}
+			if !membersToValidate[userID] {
+				membersToValidate[userID] = true
+			}
+		}
+	}
+	for userID := range membersToValidate {
+		userAPIContext := *request
+		userRequest := *(request.Request)
+		userAPIContext.Request = &userRequest
+		userAPIContext.Request.Header.Set(impersonateUserHeader, userID)
+		for projectID := range targetProjects {
+			project := &v3client.Project{}
+			if err := access.ByID(&userAPIContext, &mgmtschema.Version, v3client.ProjectType, projectID, project); err != nil {
+				request.Request.Header.Set(impersonateUserHeader, creatorID)
+				logrus.Errorf("User %v cannot access project %v", userAPIContext.Request.Header.Get(impersonateUserHeader), projectID)
```

Comment:
nit: can directly use userID here

Reply:
keeping it to ensure current request's userID is what we expected